README.md
RmarineHeatWaves
================

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/RmarineHeatWaves)](https://cran.r-project.org/package=RmarineHeatWaves) [![Travis-CI Build Status](https://travis-ci.org/ajsmit/RmarineHeatWaves.svg?branch=master)](https://travis-ci.org/ajsmit/RmarineHeatWaves) ![](https://cranlogs.r-pkg.org/badges/grand-total/RmarineHeatWaves)

The **RmarineHeatWaves** package is a translation of the original Python code written 
by Eric C. J. Oliver that can be found on [GitHub](https://github.com/ecjoliver/marineHeatWaves). 

This package is actively maintained, but on 2018-04-24 it was forked to a new package called [**heatwaveR**](https://robwschlegel.github.io/heatwaveR/index.html), which will supercede and eventually replace **RmarineHeatWaves** in the near future. For continuity, we are ensuring that the functionality in **heatwaveR** will be exactly the same in **RmarineHeatWaves**, and for a while the two packages will be maintained in sync, at least as far as developments around the existing code-base are concerned. New functions will only be added to **heatwaveR**.

The **RmarineHeatWaves** R package contains a number of functions which calculate and display marine heat waves according to the definition of Hobday et al. (2016). The marine cold spell option was implemented in version 0.13 (21 Nov 2015) of the Python module as a result of the preparation of Schlegel et al. (2017), wherein the cold events are introduced and briefly discussed.

This package may be found on [CRAN](https://cran.r-project.org/package=RmarineHeatWaves). Alternatively, you can install it from GitHub by issuing the following command:

`devtools::install_github("ajsmit/RmarineHeatWaves")`

The functions
-------------

<table style="width:47%;">
<colgroup>
<col width="30%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>detect()</code></td>
<td>The main function which detects the events as per the definition of Hobday et al. (2016).</td>
</tr>
<tr class="even">
<td><code>make_whole()</code></td>
<td>Constructs a continuous, uninterrupted time series of temperatures.</td>
</tr>
<tr class="odd">
<td><code>block_average()</code></td>
<td>Calculates annual means for event metrics.</td>
</tr>
<tr class="even">
<td><code>event_line()</code></td>
<td>Creates a line plot of marine heat waves or cold spells.</td>
</tr>
<tr class="odd">
<td><code>lolli_plot()</code></td>
<td>Creates a timeline of selected event metrics.</td>
</tr>
<tr class="even">
<td><code>exceedance()</code></td>
<td>A function similar to <code>detect()</code> but that detects consecutive days above/below a given threshold.</td>
</tr>
<tr class="odd">
<td><code>geom_flame()</code></td>
<td>Creates flame polygons of marine heat waves or cold spells.</td>
</tr>
<tr class="even">
<td><code>geom_lolli()</code></td>
<td>Creates a lolliplot timeline of selected event metric.</td>
</tr>
</tbody>
</table>

The package also provides data of observed SST records for three historical MHWs: the 2011 Western Australia event, the 2012 Northwest Atlantic event and the 2003 Mediterranean event.

The heat wave metrics
---------------------

The function will return a list of two tibbles (see the ‘tidyverse’), `clim` and `event`, which are the climatology and MHW (or MCS) events, respectively. The climatology contains the full time series of daily temperatures, as well as the the seasonal climatology, the threshold and various aspects of the events that were detected. The software was designed for detecting extreme thermal events, and the units specified below reflect that intended purpose. However, the various other kinds of extreme events may be detected according to the ‘marine heat wave’ specifications, and if that is the case, the appropriate units need to be determined by the user.

<table style="width:49%;">
<colgroup>
<col width="29%" />
<col width="19%" />
</colgroup>
<thead>
<tr class="header">
<th>Climatology metric</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>doy</code></td>
<td>Julian day (day-of-year). For non-leap years it runs 1…59 and 61…366, while leap years run 1…366. This column will be named differently if another name was specified to the <code>doy</code> argument.</td>
</tr>
<tr class="even">
<td><code>t</code></td>
<td>The date of the temperature measurement. This column will be named differently if another name was specified to the <code>x</code> argument.</td>
</tr>
<tr class="odd">
<td><code>temp</code></td>
<td>If the software was used for the purpose for which it was designed, seawater temperature [deg. C] on the specified date will be returned. This column will of course be named differently if another kind of measurement was specified to the <code>y</code> argument.</td>
</tr>
<tr class="even">
<td><code>seas_clim_year</code></td>
<td>Climatological seasonal cycle [deg. C].</td>
</tr>
<tr class="odd">
<td><code>thresh_clim_year</code></td>
<td>Seasonally varying threshold (e.g., 90th percentile) [deg. C].</td>
</tr>
<tr class="even">
<td><code>var_clim_year</code></td>
<td>Seasonally varying variance (standard deviation) [deg. C].</td>
</tr>
<tr class="odd">
<td><code>thresh_criterion</code></td>
<td>Boolean indicating if <code>temp</code> exceeds <code>thresh_clim_year</code>.</td>
</tr>
<tr class="even">
<td><code>duration_criterion</code></td>
<td>Boolean indicating whether periods of consecutive <code>thresh_criterion</code> are &gt;= <code>min_duration</code>.</td>
</tr>
<tr class="odd">
<td><code>event</code></td>
<td>Boolean indicating if all criteria that define a MHW or MCS are met.</td>
</tr>
<tr class="even">
<td><code>event_no</code></td>
<td>A sequential number indicating the ID and order of occurence of the MHWs or MCSs.</td>
</tr>
</tbody>
</table>

The events are summarised using a range of event metrics:

| Event metric   | Description                                                    |
|----------------|----------------------------------------------------------------|
| `index_start`  | Start index of event.                                          |
| `index_stop`   | Stop index of event.                                           |
| `event_no`     | A sequential number indicating the ID and order of the events. |
| `duration`     | Duration of event \[days\].                                    |
| `date_start`   | Start date of event \[date\].                                  |
| `date_stop`    | Stop date of event \[date\].                                   |
| `date_peak`    | Date of event peak \[date\].                                   |
| `int_mean`     | Mean intensity \[deg. C\].                                     |
| `int_max`      | Maximum (peak) intensity \[deg. C\].                           |
| `int_var`      | Intensity variability (standard deviation) \[deg. C\].         |
| `int_cum`      | Cumulative intensity \[deg. C x days\].                        |
| `rate_onset`   | Onset rate of event \[deg. C / day\].                          |
| `rate_decline` | Decline rate of event \[deg. C / day\].                        |

`int_max_rel_thresh`, `int_mean_rel_thresh`, `int_var_rel_thresh`, and `int_cum_rel_thresh` are as above except relative to the threshold (e.g., 90th percentile) rather than the seasonal climatology.

`int_max_abs`, `int_mean_abs`, `int_var_abs`, and `int_cum_abs` are as above except as absolute magnitudes rather than relative to the seasonal climatology or threshold.

`int_max_norm` and `int_mean_norm` are as above except units are in multiples of threshold exceedances, i.e., a value of 1.5 indicates the event intensity (relative to the climatology) was 1.5 times the value of the threshold (relative to climatology, i.e., threshold - climatology.)

Note that `rate_onset` and `rate_decline` will return `NA` when the event begins/ends on the first/last day of the time series. This may be particularly evident when the function is applied to large gridded data sets. Although the other metrics do not contain any errors and provide sensible values, please take this into account in its interpretation.

Examples of detection and graphing
----------------------------------

The `detect()` function is the package’s core function. Here is the `detect()` function applied to the Western Australian test data, which are also discussed by Hobday et al. (2016):

``` r
library(RmarineHeatWaves); library(plyr); library(dplyr); library(ggplot2)
ts <- make_whole(sst_WA)
mhw <- detect(ts, climatology_start = "1983-01-01",
              climatology_end = "2012-12-31")
mhw$event %>% 
  dplyr::ungroup() %>%
  dplyr::select(event_no, duration, date_start, date_peak, int_mean, int_max, int_cum) %>% 
  dplyr::arrange(-int_cum) %>% 
  head(5)
#> # A tibble: 5 x 7
#>   event_no duration date_start date_peak  int_mean int_max int_cum
#>      <int>    <dbl> <date>     <date>        <dbl>   <dbl>   <dbl>
#> 1       22      95. 1999-05-13 1999-05-22     2.50    3.60   237. 
#> 2       42      60. 2011-02-06 2011-02-28     3.21    6.51   193. 
#> 3       49      47. 2012-01-11 2012-01-27     2.23    3.30   105. 
#> 4       50      46. 2012-03-01 2012-04-10     1.99    2.96    91.7
#> 5       41      40. 2010-12-24 2011-01-28     2.16    3.27    86.3
```

The corresponding `event_line()` and `lolli_plot()`, which represent the massive Western Australian heatwave of 2011, look like this:

``` r
event_line(mhw, spread = 200, metric = "int_cum",
           start_date = "2010-10-01", end_date = "2011-08-30")
```

![](tools/fig-example1-1.png)

``` r
lolli_plot(mhw)
```

![](tools/fig-example2-1.png)

The `event_line()` and `lolli_plot()` functions were designed to work directly on one of the list returned by `detect()`. If more control over the figures is required, it may be useful to create them in **ggplot2** by stacking ‘geoms’. We specifically created two new **ggplot2** geoms to reproduce the functionality of `event_line()` and `lolli_plot()`. These functions are more general in their functionality and can be used outside of the **RmarineHeatWave** package too. To apply them to MHWs and MCSs, they require that we access the `clim` or `event` data frames within the list that is produced by `detect()`. Here is how:

``` r
mhw2 <- mhw$clim # find the climatology dataframe
mhw2 <- mhw2[10580:10690,] # identify the region of the time series of interest

ggplot(mhw2, aes(x = t, y = temp, y2 = thresh_clim_year)) +
  geom_flame() +
  geom_text(aes(x = as.Date("2011-02-01"), y = 28, label = "The MHW that launched\na thousand papers."))

ggplot(mhw$event, aes(x = date_start, y = int_max)) +
  geom_lolli(colour = "salmon", colour.n = "red", n = 3) +
  geom_text(aes(x = as.Date("2006-10-01"), y = 5,
                label = "The distribution of events\nis skewed towards the\nend of the time series."),
            colour = "black")
```

The default output of these function may not be to your liking. If so, not to worry. As **ggplot2** geoms, they are highly maleable. For example, if we were to choose to reproduce the format of the MHWs as seen in Hobday et al. (2016), the code would look something like this:

``` r
# It is necessary to give geom_flame() at least one row on either side of the event in order to calculate the polygon corners smoothly
mhw_top <- mhw2[49:110,]

ggplot(data = mhw2, aes(x = t)) +
  geom_flame(aes(y = temp, y2 = thresh_clim_year, fill = "all"), show.legend = T) +
  geom_flame(data = mhw_top, aes(y = temp, y2 = thresh_clim_year, fill = "top"), show.legend = T) +
  geom_line(aes(y = temp, colour = "temp")) +
  geom_line(aes(y = thresh_clim_year, colour = "thresh"), size = 1.0) +
  geom_line(aes(y = seas_clim_year, colour = "seas"), size = 1.2) +
  scale_colour_manual(name = "Line Colour",
                      values = c("temp" = "black", "thresh" =  "forestgreen", "seas" = "grey80")) +
  scale_fill_manual(name = "Event Colour", values = c("all" = "salmon", "top" = "red")) +
  guides(colour = guide_legend(override.aes = list(fill = NA))) +
  xlab("Date") +
  ylab(expression(paste("Temperature [", degree, "C]")))
```

![](tools/fig-example4-1.png)

Conversely, should we not wish to highlight any events with `geom_lolli()`, it would look like this:

``` r
# Note that this is accomplished by setting 'colour.n = NA', not by setting 'n = 0'.
ggplot(mhw$event, aes(x = date_start, y = int_cum)) +
  geom_lolli(colour = "salmon", n = 3, colour.n = NA)
```

![](tools/fig-example5-1.png)

The calculation and visualisation of marine cold spells is also accommodated within this package. Here is a cold spell detected in the OISST data for Western Australia:

``` r
mcs <- detect(ts, climatology_start = "1983-01-01", climatology_end = "2012-12-31",
              cold_spells = TRUE)
mcs$event %>% 
  dplyr::ungroup() %>%
  dplyr::select(event_no, duration, date_start, date_peak, int_mean, int_max, int_cum) %>% 
  dplyr::arrange(int_cum) %>% 
  head(5)
#> # A tibble: 5 x 7
#>   event_no duration date_start date_peak  int_mean int_max int_cum
#>      <int>    <dbl> <date>     <date>        <dbl>   <dbl>   <dbl>
#> 1       16      76. 1990-04-13 1990-05-11    -2.54   -3.22  -193. 
#> 2       54      58. 2003-12-19 2004-01-23    -1.80   -2.66  -104. 
#> 3       71      52. 2014-04-14 2014-05-05    -1.82   -2.57   -94.6
#> 4        8      38. 1986-06-24 1986-07-17    -2.01   -2.95   -76.4
#> 5       51      32. 2003-09-08 2003-09-16    -1.56   -2.12   -49.9
```

The plots showing the marine cold spells look like this:

``` r
# this function needs to be updated as it currently only works when 't' is the column name for date
event_line(mcs, spread = 200, metric = "int_cum",
           start_date = "1990-01-01", end_date = "1990-08-30")

lolli_plot(mcs)
```

![](tools/fig-example6-1.png) ![](tools/fig-example6-2.png)

Cold spell figures may be created as geoms in **ggplot2**, too:

``` r
mcs2 <- mcs$clim
mcs2 <- mcs2[2990:3190,]

# # Note that the plot centres on the polygons, so it may be necessary to manually zoom out a bit
ggplot(data = mcs2, aes(x = t)) +
  geom_flame(aes(y = thresh_clim_year, y2 = temp), fill = "steelblue3", show.legend = F) +
  geom_line(aes(y = temp, colour = "temp")) +
  geom_line(aes(y = thresh_clim_year, colour = "thresh"), size = 1.0) +
  geom_line(aes(y = seas_clim_year, colour = "seas"), size = 1.2) +
  scale_colour_manual(name = "Line Colour",
                      values = c("temp" = "black", "thresh" =  "forestgreen", "seas" = "grey80")) +
  scale_y_continuous(limits = c(18, 23.5)) +
  xlab("Date") +
  ylab(expression(paste("Temperature [", degree, "C]")))

ggplot(mcs$event, aes(x = date_start, y = int_cum)) +
  geom_lolli(colour = "steelblue3", colour.n = "navy", n = 7) +
  xlab("Date") +
  ylab(expression(paste("Cumulative intensity [days x ", degree, "C]")))
```

![](tools/fig-example7-1.png) ![](tools/fig-example7-2.png)

The exceedance function
-----------------------

In addition to the calculation of extreme events, consecutive days over a given static threshold may be calculated with the `exceedance()` function.

``` r
exc <- exceedance(ts, threshold = 25)
exc$exceedance %>%
  ungroup() %>%
  select(exceedance_no, duration, date_start, date_peak, int_mean, int_cum) %>%
  dplyr::arrange(-int_cum) %>%
  head(5)
#>   exceedance_no duration date_start  date_peak  int_mean   int_cum
#> 1             7       52 2011-02-08 2011-02-28 1.6740379 87.049969
#> 2             6       25 2008-04-03 2008-04-14 0.9799994 24.499985
#> 3            10       41 2012-03-03 2012-04-10 0.4385360 17.979977
#> 4             2       17 1999-05-13 1999-05-22 0.8558818 14.549990
#> 5             5       10 2000-05-03 2000-05-04 0.6969994  6.969994
```

The same function may be used to calculate consecutive days below a threshold, too.

``` r
exc <- exceedance(ts, threshold = 19, below = TRUE)
exc$exceedance %>%
  dplyr::ungroup() %>%
  dplyr::select(exceedance_no, duration, date_start, date_peak, int_mean, int_cum) %>%
  dplyr::arrange(int_cum) %>%
  head(5)
#>   exceedance_no duration date_start  date_peak   int_mean   int_cum
#> 1            17       46 2003-09-06 2003-09-16 -0.6008700 -27.64002
#> 2            16       31 2002-09-08 2002-09-25 -0.8480649 -26.29001
#> 3            13       24 1997-09-03 1997-09-15 -0.7691671 -18.46001
#> 4            20       25 2005-09-26 2005-10-12 -0.5420004 -13.55001
#> 5            12       18 1997-08-13 1997-08-22 -0.6944449 -12.50001
```

Working with gridded SST data
-----------------------------


NEWS.md
# VERSION CHANGES

## Changes in version 0.17.0 (15 October 2018)
* Minor bug fix in RmarineHeatWaves.R, i.e. sort function in 'clim_only'
  on l.336.

## Changes in version 0.17.0 (4 June 2018)
* Add start-up message.
* Minor bug fixes.
* Reached end of active development cycle.

## Changes in version 0.16.4 (28 April 2018)
* Remove restriction to require full years for start/end points of
  climatology calculations in detect().
* Documentation updated accordingly.
* Fix error with smooth_percentile and smooth_percentile_width descriptions
  that were interchanged.

## Changes in version 0.16.3 (18 April 2018)
* Bug fix to the block_event() function.

## Changes in version 0.16.2 (3 March 2018)
* Bug fix to the event_line() function.

## Changes in version 0.16.1 (28 February 2018)
* Repaired event_line(), which broke in v.0.16.0 due to an edit there
  to enable using arbitrary column headers for the date and temperature columns
  (previously the defaults were 't' and 'temp').

## Changes in version 0.16.0.9000 (18 January 2018)
* detect() now reurns a list with two tibbles, rather than data.frames as before.

## Changes in version 0.16.0 (6 January 2018)
* The make_whole(), detect() and exceedance() functions now handle any arbitrary
  column names, which are specified via arguments to the function.
* Remove dependence on magrittr.
* More consistent use of tidyverse packages throughout.
* Updates to examples.
* Miscellaneous other improvements troughout the function internals.

## Changes in version 0.15.9 (19 June 2017)
* Warnings given by exceedance() when exceedances begin/end on the first/last day 
  of the time series have been corrected --- it will now report an NA when the 
  exceedance occurs at the start/end of the time series.
* Some plyr routines phased out in exceedance function in favour of a dplyr approach.
  
## Changes in version 0.15.8 (18 June 2017)
* Further removed need for plyr functions --- use dplyr instead.

## Changes in version 0.15.7 (18 June 2017)
* Warnings given by detect() when events begin/end on the first/last day of the
  time series have been corrected --- it will now report an NA when the event
  occurs at the start/end of the time series. This bug was highlighted by Juan Bazo.
* Documentation in detect() about the use of percentiles with cold-spells has 
  been made more clear.
* Dependency on some plyr functions have been started to be phased out; this is 
  done to eventually ensure full compliance with the tidyverse suite of functions.
* Remove some unwanted attributes that resulted from dplyr::group_by on lines 496
  and 497. The list containing 'event' and 'clim' is now uncluttered by unwanted
  attributes.

## Changes in version 0.15.6 (6 March 2017)
* Update README.Rmd to correct figure paths.
 
## Changes in version 0.15.5 (6 March 2017)
* Update to 'protoFunc' helper function within 'detect'; in the previous version
    it returned the error "Error: not compatible with STRSXP" under rare circumstances,
    the cause of which remains unknown.
* Update the calculation of proto_gaps for the same reason.

## Changes in version 0.15.4 (5 March 2017)
* The detect() function now also produces a 'climatological' variance.
* Improve documentation in a help file.

## Changes in version 0.15.3 (28 February 2017)
* The detect() function may now be told to calculate climatologies only.

## Changes in version 0.15.2 (18 February 2017)
* Updating README.md, README.Rmd files.
* Updating the vignette.

## Changes in version 0.15.1 (15 February 2017)
* 'exceedence' changed to 'exceedance' throughout.

## Changes in version 0.15.0 (15 February 2017)
* Final edits to 'geom_flame' and 'geom_lolli';
  - specifically in geom_flame, small gaps in the corners of polygons are now 
  being filled in correctly.
* Updated examples.

## Changes in version 0.14.6 (15 February 2017)
* Edits to 'geom_flame' and 'geom_lolli'.

## Changes in version 0.14.5 (14 February 2017)
* Edits to 'geom_flame' and 'geom_lolli'.

## Changes in version 0.14.4 (11 February 2017)
* Changed 'geom_event_line' and 'geom_lolli_plot' to 'geom_flame' and 'geom_lolli'
* These new functions are now proper ggplot geoms.

## Changes in version 0.14.3 (31 January 2017)
* Corrected plotting error in event polygons in 'event_line' function
* Added two new functions: 'geom_event_line' and 'geom_lolli_plot' which may be 
  used with ggplot2 directly as geoms.

## Changes in version 0.14.2 (26 January 2017)
* Fix bug which was that was brought to my attention by Mahmoud Haouari: enable 
  the pctile option in the detect() function, which was accidentally disabled.

## Changes in version 0.14.1 (10 January 2017)
* Improved error messages in exceedance().
* Bug fix.

## Changes in version 0.14.0 (11 December 2016)
* All changes since version 0.13.1.

## Changes in version 0.13.9.9200 (11 December 2016)
* Minor edits to documentation.
* Corrected issue where the event_line graph's y-axis did not reflect the correct 
  units for the metric selected for plotting.
* Updated the theme, i.e. removed the default ggplot2 theme in favour
  of a more conventional white background with nice black axes.
* Removed the legend that indicated the peak and secondary events from the 
  event_line plot. 
* Repositioned the legend in the event_line plot: it is now int he bottom right 
  corner as here it has less chance of plotting over the peak events when the 
  graph is rescaled.

## Changes in version 0.13.8.9200 (23 November 2016)
* Minor update to exceedance() function to improve usability.
* Usage of exceedance() shown in README file

## Changes in version 0.13.7.9100 (22 October 2016)
* The additions to the detect() function madde in the previous update have been 
  rolled back in favour of creating an independent function to calculate exceedances.

## Changes in version 0.13.6.9000 (22 October 2016)
* Added 'threshold' variable to detect() function.

R/ggplotGeoms.R
#' Create flame ploygons of Marine Heat Waves or Cold Spells.
#'
#' Creates a ggplot2 geom of warm or cold events as per the second row of Figure 3 in
#' Hobday et al. (2016).
#'
#' @inheritParams ggplot2::layer
#' @param @param na.rm If \code{FALSE} (the default), removes missing values with
#'    a warning.  If \code{TRUE} silently removes missing values.
#' @param ... other arguments passed on to \code{\link{layer}}. These are
#'   often aesthetics, used to set an aesthetic to a fixed value, like
#'   \code{color = "red"} or \code{size = 3}. They may also be parameters
#'   to the paired geom/stat.
#' @param stat.top Choose which statistic to use to highlight the top event
#' found within the time series. Option are: \code{int_max} for the maximum 
#' intensity of the event, \code{int_cum} for the cummulative intensity of 
#' the event, \code{int_mean} for the mean intensity of the event and \code{dur}
#' for the duration of the event.
#'
#' @return The function will create a ggplot2 geom indicating the climatology,
#' threshold and temperature, with the hot or cold events that meet the
#' specifications of Hobday et al. (2016) shown as flame polygons. The function
#' automagically detects if one is plotting MHWs or MCS and adjusts the colour 
#' palette accordingly. These default colours may still be overridden as normal.
#' The top event detect during the selected time period, as determined by the
#' selected statistics given in \code{stat.top} will be visible in a
#' brighter colour. The default colour may be overidden in the aesthetics as well. 
#' This function differs in use from \code{\link{event_line}}
#' in that it must be created as a ggplot() object. The benefit of this being
#' that one may add additional information to the figure as may be necessary.
#'
#' @author Robert W. Schlegel
#'
#' @references Hobday, A.J. et al. (2016), A hierarchical approach to defining
#' marine heatwaves, Progress in Oceanography, 141, pp. 227-238,
#' doi: 10.1016/j.pocean.2015.12.014
#'
#' @export
#'
#' @examples
#' t_dat <- make_whole(sst_WA)
#' res <- detect(t_dat, climatology_start = 1983, climatology_end = 2012) # using default values
#' mhw <- res$clim
#' mhw <- mhw[10580:10690,]
#'
#' \dontrun{
#' library(ggplot2)
#' ggplot(mhw, aes(x = date, y = temp, thresh = thresh_clim_year, seas = seas_clim_year, event = event_no)) +
#'  geom_flame() +
#'  geom_text(aes(x = as.Date("2011-02-01"), y = 28, label = "Wow. Such heatwave. Many warm."))
#' }

geom_flame <- function(mapping = NULL, data = NULL,
                       stat = "identity", position = "identity",
                       ...,
                       stat.top = "int_cum",
                       na.rm = FALSE,
                       show.legend = NA,
                       inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomFlameOn,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      stat.top = stat.top,
      na.rm = na.rm,
      ...
    )
  )
}

GeomFlame <- ggproto("GeomFlame", Geom,
                     
                     required_aes = c("x", "y", "thresh"),
                     
                     default_aes = aes(colour = NA, fill = NA,
                                       size = 0.5, linetype = 1, alpha = NA),
                     
                     draw_key = draw_key_polygon,
                     
                     draw_group = function(data, panel_scales, coord, na.rm = FALSE) {
                       if (na.rm) data <- data[stats::complete.cases(data[c("x", "y", "thesh")]), ]
                       
                       # Check that aesthetics are constant
                       aes <- unique(data[c("colour", "fill", "size", "linetype", "alpha")])
                       if (nrow(aes) > 1) {
                         stop("Aesthetics must be consistent")
                       }
                       
                       aes <- as.list(aes)
                       
                       missing_pos <- !stats::complete.cases(data[c("x", "y", "thresh")])
                       ids <- cumsum(missing_pos) + 1
                       ids[missing_pos] <- NA
                       
                       positions <- plyr::summarise(data,
                                                    x = c(x, rev(x)), y = c(y, rev(thresh)), id = c(ids, rev(ids)))
                       munched <- coord_munch(coord, positions, panel_scales)
                       
                       grid::polygonGrob(
                         munched$x, munched$y, id = munched$id,
                         default.units = "native",
                         gp = grid::gpar(
                           fill = alpha(aes$fill, aes$alpha),
                           col = aes$colour,
                           lwd = aes$size * .pt,
                           lty = aes$linetype)
                       )
                     }
)

GeomFlameOn <- ggproto("GeomFlameOn", Geom,
                       required_aes = c("x", "y", "thresh", "seas"), # "event" intentionally not included here as a required aes, but it does require it...
                       
                       default_aes = aes(colour = NA, fill.MHW = "salmon", fill.MHW.top = "red", 
                                         fill.MCS = "steelblue3", fill.MCS.top = "navy", 
                                         size = 0.5, linetype = 1, alpha = NA,
                                         temp.line = "black", seas.line = "grey60", thresh.line = "forestgreen"),
                       
                       draw_key = draw_key_polygon,
                       
                       draw_group = function(data, panel_scales, coord, stat.top, na.rm = FALSE){
                         
                         # Deduce if MHWs or MCSs are being measured and set default fill colours accordingly
                         flame_all <- data
                         if(flame_all$y[complete.cases(flame_all$event)][1] > flame_all$thresh[complete.cases(flame_all$event)][1]){
                           flame_all$y[flame_all$y < flame_all$thresh] <- NA
                           flame_top <- flame_all
                           flame_all$fill <- data$fill.MHW
                           flame_top$fill <- data$fill.MHW.top
                         } else if (flame_all$y[complete.cases(flame_all$event)][1] < flame_all$thresh[complete.cases(flame_all$event)][1]){
                           flame_all$y[flame_all$y > flame_all$thresh] <- NA
                           flame_top <- flame_all
                           flame_all$fill <- data$fill.MCS
                           flame_top$fill <- data$fill.MCS.top
                         }
                         
                         # Choose event to subset
                         flame_stats <- plyr::ddply(data[complete.cases(data$event),], "event", plyr::summarise,
                                                    int_max = max(abs(y-thresh),na.rm = T),
                                                    int_cum = abs(sum(y-thresh, na.rm = T)),
                                                    int_mean = abs(mean(y-thresh, na.rm = T)),
                                                    dur = length(y))
                         flame_var <- data.frame(event = flame_stats$event, var = flame_stats[,colnames(flame_stats) == as.character(stat.top)])
                         flame_top <- flame_top[flame_top$event == flame_var$event[flame_var$var == max(flame_var$var)],]
                         
                         # Create temperature line info
                         temp_line <- data
                         temp_line$colour <- data$temp.line
                         
                         # Create seasonal climatology line info
                         seas_line <- data
                         seas_line$y <- data$seas
                         seas_line$colour <- data$seas.line
                         
                         # Create threshold line info
                         thresh_line <- data
                         thresh_line$y <- data$thresh
                         thresh_line$colour <- data$thresh.line
                         
                         # The list of geoms to create
                         grid::gList(
                           GeomFlame$draw_panel(flame_all, panel_scales, coord),
                           GeomFlame$draw_panel(flame_top, panel_scales, coord),
                           ggplot2::GeomLine$draw_panel(temp_line, panel_scales, coord),
                           ggplot2::GeomLine$draw_panel(seas_line, panel_scales, coord),
                           ggplot2::GeomLine$draw_panel(thresh_line, panel_scales, coord)
                         )
                       }
)

#' Create a Timeline geom of a selected event metric.
#'
#' Visualise a timeline of several event metrics as 'lollipops'.
#' 
#' @inheritParams ggplot2::layer
#' @param @param na.rm If \code{FALSE} (the default), removes missing values with
#'    a warning.  If \code{TRUE} silently removes missing values.
#' @param ... other arguments passed on to \code{\link{layer}}. These are
#'   often aesthetics, used to set an aesthetic to a fixed value, like
#'   \code{color = "red"} or \code{size = 3}. They may also be parameters
#'   to the paired geom/stat.
#' @param n The number of top events to highlight. Default is 1.
#'
#' @return The function will return a graph of the intensity of the selected
#' metric along the y-axis versus a time variable along the x axis.
#' The number of top events (\code{n}) from the chosen metric may be highlighted
#' in a brighter colour with the aesthetic value \code{coulour.n}. 
#' This function differs in use from \code{\link{lolli_plot}}
#' in that it must be created as a ggplot object. The benefit of this being
#' that one may add additional information to the figure as may be necessary.
#'
#' @author Robert W. Schlegel
#'
#' @export
#'
#' @examples
#' t_dat <- make_whole(sst_NW_Atl)
#' res <- detect(t_dat, climatology_start = 1983, climatology_end = 2012) # using default values
#'
#' \dontrun{
#' ggplot() + geom_lolli_plot(res, metric = "int_cum", event_count = 3, xaxis = "date_peak") +
#' geom_text(aes(x = as.Date("2007-01-01"), y = 375,
#' label = "One may clearly see\njust how dramatic\n these heatwaves are."))
#' }

geom_lolli<- function(mapping = NULL, data = NULL, 
                      ...,
                      n = 1,
                      na.rm = FALSE, show.legend = NA, inherit.aes = TRUE) {
  
  layer(
    data = data,
    mapping = mapping,
    stat = "identity",
    geom = GeomLolli,
    position = "identity",
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      n = n,
      ...
    )
  )
}

GeomLolli <- ggproto("GeomLolli", Geom,
                     required_aes = c("x", "y"),
                     default_aes = aes(shape = 19, colour = "grey30", size = 1, fill = NA,
                                       alpha = NA, stroke = 1, colour.n = "black"),
                     
                     draw_key = draw_key_point,
                     
                     draw_group = function(data, panel_scales, coord, n) {
                       data$xend = data$x 
                       data$yend = 0
                       data = data[order(abs(data$y), decreasing = T),]
                       
                       # Define the big points
                       big_points = data
                       big_points$size = data$size*2
                       
                       # Define the top n events
                       data_n = data[1:n,]
                       data_n$colour = data$colour.n[1:n]
                       big_points_n = big_points[1:n,]
                       big_points_n$colour = data$colour.n[1:n]
                       
                       # Define the look of the small white fillings
                       small_points = data
                       small_points$size = data$size/2
                       small_points$colour = "white"
                       
                       grid::gList(
                         ggplot2::GeomSegment$draw_panel(data, panel_scales, coord),
                         ggplot2::GeomPoint$draw_panel(big_points, panel_scales, coord),
                         ggplot2::GeomSegment$draw_panel(data_n, panel_scales, coord),
                         ggplot2::GeomPoint$draw_panel(big_points_n, panel_scales, coord),
                         ggplot2::GeomPoint$draw_panel(small_points, panel_scales, coord)
                       )
                       
                     }
)
